### ISSUE TITLE
Shell not executing bash

### ISSUE THREAD

**Issue by dholt** [apptainer/singularity#194]
After updating from 2.0 to 2.1.2 the behavior when executing a shell has changed. I no longer end up in a bash shell, and when I execute bash it clobbers my prompt (not a big deal but it's a nice reminder that I'm in a different namespace). Is there possibly something in my environment conflicting with singularity?

`[root@cluster ~]# singularity shell -w ubuntu.img`
`Singularity.ubuntu.img> # echo $SHELL`
`/bin/sh`
`Singularity.ubuntu.img> # bash`
`root@cluster:~#`

**Comment by gmkurtzer** [comment]
Hi Doug,

There are a couple of things that maybe going on. Is this a new container that was created with 2.1.2, or was it created with 2.0? If it was created with 2.1.2, then you should find that there is a file /.shell which will be called when you execute "singularity shell...". Inside that file you will see that it should opt for using `$SHELL` if it is defined. If that file does not exist it will call /bin/sh. 

In either case... Bash will execute the login scripts and dot files which clobber the Singularity defined PS1 prompt. If you call Bash with --noprofile and --norc, it shouldn't clobber it automatically.

BTW, if you want to add those files (/.shell, /.exec, /.run) to your container, you can re-bootstrap your container with no boostrap file: `sudo singularity bootstrap ubuntu.img`

Hope that helps!

Greg

**Comment by dholt** [comment]
Thanks Greg,

I tried both with the older container created with 2.0 and one created with 2.1.2. When launching the newer container I do have `$SHELL` defined but it still launches `/bin/sh`:

`dholt@host:~$ echo $SHELL`
`/bin/bash`
`dholt@host:~$ singularity shell ubuntu.img`
`Singularity.ubuntu.img> $ echo $SHELL`
`/bin/sh`
`Singularity.ubuntu.img> $ ls -l /bin/bash`
`-rwxr-xr-x 1 root root 1017016 Apr  9  2014 /bin/bash`

**Comment by gmkurtzer** [comment]
Can you confirm that you have /.shell within the container?

**Comment by dholt** [comment]
Yes (sorry it doesn't format well):

`dholt@host:~$ singularity exec ubuntu.img cat /.shell
#!/bin/sh
. /environment
if test -n "$SHELL" -a -x "$SHELL"; then
    exec "$SHELL" "$@"
else
    echo "ERROR: Shell does not exist in container: $SHELL" 1>&2
    echo "ERROR: Using /bin/sh instead..." 1>&2
fi
if test -x /bin/sh; then
    SHELL=/bin/sh
    export SHELL
    exec /bin/sh "$@"
else
    echo "ERROR: /bin/sh does not exist in container" 1>&2
fi
exit 1`

**Comment by gmkurtzer** [comment]
Ohhhh, .... I see what's happening. Sorry, I'm juggling the development version in my head and it was confusing me. SHELL is overridden by libexec/cli/shell.exec to use /bin/sh unless it is specified as something different via the -s/--shell option. Try:

```
$ singularity shell --shell /bin/bash ubuntu.img --noprofile --norc
```

**Comment by dholt** [comment]
That seems to work, thanks! I'm guessing the way to set that behavior as default is just to modify /.shell in the container?

**Comment by gmkurtzer** [comment]
Best to modify /environment and define SHELL=/bin/bash from there...

I think for the future, it should use SINGULARITY_SHELL from the environment and use that to overwrite SHELL only if defined. This way you can easily set that outside your container and always have the appropriate environment inside the container.

**Comment by cfriedline** [comment]
This may also depend on the Ubuntu version, since they symlink `/bin/sh -> dash`, rather than `bash`, in recent iterations. You can get around this in your setup by doing `dkpg-reconfigure` dash in the container, choosing `No` as the default shell.  Or change the `environment` like @gmkurtzer says.

**Comment by czka** [comment]
Ubuntu's /bin/dash is not an issue here. I have same symptoms with Singularity 2.1.2 on a newly built Debian image:

```
$ singularity shell /tmp/Debian.img
Singularity.Debian.img> $ echo $SHELL
/bin/sh
```

vs:

```
$ singularity shell -s /bin/bash /tmp/Debian.img --norc --noprofile
Singularity.Debian.img> $ echo $SHELL
/bin/bash
```

```
Singularity.Debian.img> $ cat /.shell   
#!/bin/sh
. /environment
if test -n "$SHELL" -a -x "$SHELL"; then
    exec "$SHELL" "$@"
else
    echo "ERROR: Shell does not exist in container: $SHELL" 1>&2
    echo "ERROR: Using /bin/sh instead..." 1>&2
fi
if test -x /bin/sh; then
    SHELL=/bin/sh
    export SHELL
    exec /bin/sh "$@"
else
    echo "ERROR: /bin/sh does not exist in container" 1>&2
fi
exit 1
```

@gmkurtzer 
http://singularity.lbl.gov/#cli_shell reads `default shell that it will attempt to call is /bin/bash`. Looks like the online docs are not in sync with the `singularity help shell` output, which reads `Obtain a shell (/bin/sh)`.

**Comment by gmkurtzer** [comment]
Good catch @czka, thank you! Fixed in [gh-pages 4a9dc3b].

**Comment by gmkurtzer** [comment]
Closing the ticket as I think it is resolved. If not, feel free to comment on it and I'll reopen.

Thanks!
